From d7029cd8ecd4adc212864c676f9e9457c8cb5db1 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Tue, 2 Feb 2016 14:12:48 +0530
Subject: [PATCH 11/13] [EM/PMIC_CCSM]: Fix usb enumeration issue during data
 role swap

When doing data role swap between two devices, sometimes device
enumeration is getting failed in host side. This is due to the
SOC asserts the reset signal to the USB PHY. Hence do not switch
to soc to assert reset signal to usb phy during data role swap
unless actual connect happends

Change-Id: Iaa638e4d3d4984363cdff8d722f39862f395c49e
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-18545
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
Reviewed-on: https://android.intel.com:443/465420
---
 drivers/power/intel_pmic_ccsm.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/drivers/power/intel_pmic_ccsm.c b/drivers/power/intel_pmic_ccsm.c
index e9c2c5e..52453cf 100644
--- a/drivers/power/intel_pmic_ccsm.c
+++ b/drivers/power/intel_pmic_ccsm.c
@@ -81,6 +81,11 @@
 #define DECI_KELVIN_TO_CELSIUS(t) ((t - KELVIN_OFFSET) / 10)
 #define CELSIUS_TO_DECI_KELVIN(t) (((t * 100) + KELVIN_OFFSET) / 10)
 
+#define IS_ALL_CABLE_DISCONNECTED(x)	(!(x)->host_cable_state &&	\
+					!(x)->device_cable_state &&	\
+					!(x)->snk_cable_state &&	\
+					!(x)->src_cable_state)
+
 /* Type definitions */
 static void pmic_bat_zone_changed(void);
 static int intel_pmic_handle_otgmode(bool enable);
@@ -2033,8 +2038,12 @@ static void pmic_ccsm_process_cable_events(enum cable_type cbl_type,
 	if (notify_otg) {
 		dev_dbg(chc.dev, "%s notified %d to otg\n", __func__, otg_evt);
 		atomic_notifier_call_chain(&chc.otg->notifier, otg_evt, NULL);
-		pmic_write_reg(chc.reg_map->pmic_usbphyctrl,
-				cable_state ? USBPHYRSTB_EN : USBPHYRSTB_DIS);
+		if (IS_ALL_CABLE_DISCONNECTED(&chc))
+			pmic_write_reg(chc.reg_map->pmic_usbphyctrl,
+					USBPHYRSTB_DIS);
+		else if (cable_state)
+			pmic_write_reg(chc.reg_map->pmic_usbphyctrl,
+					USBPHYRSTB_EN);
 	}
 
 vbus_fail:
-- 
1.9.1

